src/virtio/block.c

```diff
@@ -0,0 +1,463 @@
+}
+
+static void virtio_blk_used_buffer_virq_inject(struct virtio_device *dev)
+{
+    // set the reason of the irq: used buffer notification to virtio
+    dev->data.InterruptStatus = BIT_LOW(0);
+
+    bool success = virq_inject(GUEST_VCPU_ID, dev->virq);
+    assert(success);
+}
+
+/* Set response to virtio request to error */
+static void virtio_blk_set_req_fail(struct virtio_device *dev, uint16_t desc)
+{
+    struct virtq *virtq = &dev->vqs[VIRTIO_BLK_DEFAULT_VIRTQ].virtq;
+
+    uint16_t curr_virtio_desc = desc;
+    for (;virtq->desc[curr_virtio_desc].flags & VIRTQ_DESC_F_NEXT; curr_virtio_desc = virtq->desc[curr_virtio_desc].next){}
+    *((uint8_t *)virtq->desc[curr_virtio_desc].addr) = VIRTIO_BLK_S_IOERR;
+}
+
+static void virtio_blk_set_req_success(struct virtio_device *dev, uint16_t desc)
+{
+    struct virtq *virtq = &dev->vqs[VIRTIO_BLK_DEFAULT_VIRTQ].virtq;
+
+    uint16_t curr_virtio_desc = desc;
+    for (;virtq->desc[curr_virtio_desc].flags & VIRTQ_DESC_F_NEXT; curr_virtio_desc = virtq->desc[curr_virtio_desc].next){}
+    *((uint8_t *)virtq->desc[curr_virtio_desc].addr) = VIRTIO_BLK_S_OK;
+}
+
+static bool sddf_make_req_check(blk_queue_handle_t *h, uint16_t sddf_count) {
+    // Check if req store is full, if data region is full, if req queue is full
+    // If these all pass then this request can be handled successfully
+    if (datastore_full(&datastore)) {
+        LOG_BLOCK_ERR("Request store is full\n");
+        return false;
+    }
+
+    if (blk_req_queue_full(h)) {
+        LOG_BLOCK_ERR("Request queue is full\n");
+        return false;
+    }
+
+    if (fsmem_full(&fsmem_data, sddf_count)) {
+        LOG_BLOCK_ERR("Data region is full\n");
+        return false;
+    }
+
+    return true;
+}
+
+static int virtio_blk_mmio_queue_notify(struct virtio_device *dev)
+{
+    // @ericc: If multiqueue feature bit negotiated, should read which queue from dev->QueueNotify,
+    // but for now we just assume it's the one and only default queue
+    virtio_queue_handler_t *vq = &dev->vqs[VIRTIO_BLK_DEFAULT_VIRTQ];
+    struct virtq *virtq = &vq->virtq;
+
+    blk_queue_handle_t *queue_handle = dev->sddf_handlers[SDDF_BLK_DEFAULT_HANDLE].queue_h;
+
+    bool has_dropped = false; /* if any request has to be dropped due to any number of reasons, this becomes true */
+    
+    /* get next available request to be handled */
+    uint16_t idx = vq->last_idx;
+
+    LOG_BLOCK("------------- Driver notified device -------------\n");
+    for (;idx != virtq->avail->idx; idx++) {
+        uint16_t desc_head = virtq->avail->ring[idx % virtq->num];
+
+        uint16_t curr_desc_head = desc_head;
+
+        // Print out what the request type is
+        struct virtio_blk_outhdr *virtio_req = (void *)virtq->desc[curr_desc_head].addr;
+        LOG_BLOCK("----- Request type is 0x%x -----\n", virtio_req->type);
+        
+        // Parse different requests
+        switch (virtio_req->type) {
+            // header -> body -> reply
```

Comment:
A little extra goes a long way once we've forgotten everything:

```suggestion
            /* There are three parts with each block request. The header, body (which contains the data) and reply. */
```